File: archive_ts/src/core/GlobalDomain.ts
import { TPositionMap, Token, TypeArray, TypeSet, Word } from "./Types.ts";

export interface IGlobalDomain {
    gws: TypeSet<Word>;

    gwa: TypeArray<Word>;

    wpm: TPositionMap<Word>;

    toTokens(wa: TypeArray<Word>): TypeArray<Token>;

    toWords(ta: TypeArray<Token>): TypeArray<Word>;
}

export class GlobalDomain implements IGlobalDomain {
    gws: TypeSet<Word>;

    gwa: TypeArray<Word>;

    wpm: TPositionMap<Word> = new Map();

    constructor(gws: TypeSet<Word>) {
        this.gws = gws;

        this.gwa = [...gws].sort();

        const { gwa, wpm } = this;

        for (let i = 0; i < gwa.length; i++) {
            wpm.set(gwa[i], i);
        }
    }

    toTokens = (wa: TypeArray<Word>): TypeArray<Token> => wa.map((w) => this.wpm.get(w)!);

    toWords = (ta: TypeArray<Token>): TypeArray<Word> => ta.map((t) => this.gwa[t]);
}

File: src/core/global_domain.py
from typing import List, Set
from core.types import TPositionMap


class GlobalDomain:
    words: Set[str]

    values: List[str]

    wordPositionMap: TPositionMap[str]

    def __init__(self, words: Set[str]) -> None:
        self.words = words

        self.values = list(words)  # TODO: sort

        self.wordPositionMap = dict()

        for i, value in enumerate(self.values):
            self.wordPositionMap[value] = i

    def toWords(self, tokens: List[int]) -> List[str]:
        return list(map(lambda t: self.values[t], tokens))

    def toTokens(self, words: List[str]) -> List[int]:
        return list(map(lambda w: self.wordPositionMap[w], words))
